File: otg-touch-screen/touch-to-hidg.c
/*
 * touch-to-hidg.c - 板载触摸屏 event0 -> USB HID gadget hidg0
 *
 * Build: make  或 arm-xxx-gcc touch-to-hidg.c -o touch-to-hidg
 * Run:   ./touch-to-hidg
 * Env:   TOUCH_MODE=single|multi (默认 single)
 *
 * ============================================================================
 * 数据转换总览（event0 -> hidg0）
 * ============================================================================
 *
 *  [Goodix 触摸芯片]
 *        |
 *        v
 *  Linux input 子系统 (/dev/input/event0)
 *        |  多条 struct input_event（不能直接把 event 原样 write 到 hidg）
 *        v
 *  本程序：先“攒一帧”，再组 USB HID 输入报告
 *        |
 *        v
 *  /dev/hidg0  (长度必须 == report_length，由 usb-gadget-touch.sh 配置)
 *        |
 *        v
 *  USB 主机 (Windows/Linux) 按 report_desc 解析为触摸屏
 *
 * ----------------------------------------------------------------------------
 * 第一步：从 event0 读出什么？
 * ----------------------------------------------------------------------------
 * 每条 input_event 含: type, code, value（时间戳本程序忽略）
 *
 *   type=EV_ABS  更新坐标/多点槽位:
 *     ABS_MT_SLOT         value = 当前手指槽号 0~4
 *     ABS_MT_TRACKING_ID  value>=0 按下, -1 抬起
 *     ABS_MT_POSITION_X/Y  该槽位的本地坐标 (Goodix: X 0~1023, Y 0~599)
 *     ABS_X / ABS_Y       部分驱动额外发的单点坐标（备份到 slot[0]）
 *
 *   type=EV_KEY  BTN_TOUCH  value=1 按下 / 0 抬起（单点辅助）
 *
 *   type=EV_SYN  SYN_REPORT  表示“本帧事件结束”，此前 ABS/KEY 同属一帧
 *
 * 本程序在收到 SYN_REPORT 之前只更新内存中的 slots[]，不写 hidg。
 *
 * ----------------------------------------------------------------------------
 * 第二步：内存状态 slots[5]（本地坐标系）
 * ----------------------------------------------------------------------------
 *   slots[i].active  <- TRACKING_ID >= 0
 *   slots[i].x/y     <- POSITION_X/Y（单位：板子屏 0~1023 x 0~599）
 *
 * 单点模式：build_report 只取第一个 active 的 slot；tip 还可参考 BTN_TOUCH。
 *
 * ----------------------------------------------------------------------------
 * 第三步：坐标映射（本地 -> HID 逻辑坐标）
 * ----------------------------------------------------------------------------
 *   hx = x * HID_X_MAX / SRC_X_MAX   (默认 4095/1023)
 *   hy = y * HID_Y_MAX / SRC_Y_MAX   (默认 4095/599)
 * 可选 INVERT_Y: hy = HID_Y_MAX - hy
 *
 * ----------------------------------------------------------------------------
 * 第四步：组装 HID 报告（必须与 usb-gadget-touch.sh 里 report_desc 一致）
 * ----------------------------------------------------------------------------
 *
 * A) report_length == 8（默认单点）
 *    byte0  Report ID = 0x01
 *    byte1  Tip Switch (按下 0x01)
 *    byte2  Contact ID = 0
 *    byte3-4 X uint16 小端
 *    byte5-6 Y uint16 小端
 *    byte7  Contact Count (按下 1 / 抬起 0)
 *
 * B) report_length == 64（TOUCH_MODE=multi）
 *    每指 6 字节 x 5 + byte30=contact count，其余填 0
 *
 * ----------------------------------------------------------------------------
 * 第五步：write(hidg0)
 * ----------------------------------------------------------------------------
 *   - 每次 write 恰好 report_length 字节
 *   - 与上一帧完全相同则跳过（减压 dwc2）
 *   - 主机 UDC state 须为 configured
 *
 * 可用 evtest /dev/input/event0 对照本地事件，用 test-hidg.py 对照 USB 输出。
 * ============================================================================
 */

#define _GNU_SOURCE

#include <errno.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#include <linux/input.h>
#include <poll.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#define INPUT_DEV_DEFAULT   "/dev/input/event0"
#define HIDG_DEV_DEFAULT    "/dev/hidg0"
#define GADGET_DIR          "/sys/kernel/config/usb_gadget/touch"
#define REPORT_LEN_PATH     GADGET_DIR "/functions/hid.usb0/report_length"
#define GADGET_UDC_PATH     GADGET_DIR "/UDC"

#define NUM_SLOTS           5
#define SRC_X_MAX           1023
#define SRC_Y_MAX           599
#define HID_X_MAX           4095
#define HID_Y_MAX           4095

#define HOST_WAIT_SEC       15
#define MIN_WRITE_INTERVAL_NS 8000000LL  /* 8 ms */
#define POLL_TIMEOUT_MS     1000

#ifndef INVERT_Y
#define INVERT_Y            0
#endif

#ifndef GRAB_INPUT
#define GRAB_INPUT          1
#endif

#define REPORT_MAX          64

/* linux/input-event-codes.h may be missing on old toolchain */
#ifndef ABS_MT_SLOT
#define ABS_MT_SLOT         0x2f
#endif
#ifndef ABS_MT_TRACKING_ID
#define ABS_MT_TRACKING_ID  0x39
#endif
#ifndef ABS_MT_POSITION_X
#define ABS_MT_POSITION_X   0x35
#endif
#ifndef ABS_MT_POSITION_Y
#define ABS_MT_POSITION_Y   0x36
#endif
#ifndef BTN_TOUCH
#define BTN_TOUCH           0x14a
#endif

/* 一个触摸点的中间状态（仍为板子本地坐标，尚未映射到 HID） */
struct slot {
	bool active;	/* TRACKING_ID>=0 或单点逻辑上的“有手指” */
	int x;		/* ABS_MT_POSITION_X，范围约 0~SRC_X_MAX */
	int y;		/* ABS_MT_POSITION_Y，范围约 0~SRC_Y_MAX */
};

struct app {
	int ev_fd;
	int hid_fd;
	int report_len;
	bool multi_mode;
	bool btn_touch;
	int cur_slot;
	struct slot slots[NUM_SLOTS];
	uint8_t last_report[REPORT_MAX];
	bool have_last;
	struct timespec last_write_ts;
};

/*
 * 坐标线性映射: 板子 (0..src_max) -> HID 描述符逻辑值 (0..dst_max)
 * 例: x=512, SRC_X_MAX=1023, HID_X_MAX=4095 -> hx=2048
 */
static int scale_coord(int v, int src_max, int dst_max)
{
	if (v < 0)
		v = 0;
	if (v > src_max)
		v = src_max;
	return (int)((long long)v * dst_max / src_max);
}

static int timespec_cmp_now(const struct timespec *ts, long long add_ns)
{
	struct timespec now;

	clock_gettime(CLOCK_MONOTONIC, &now);
	if (now.tv_sec > ts->tv_sec + 1)
		return 1;
	if (now.tv_sec < ts->tv_sec)
		return 0;
	{
		long long nsec = (long long)(now.tv_sec - ts->tv_sec) * 1000000000LL
			+ (now.tv_nsec - ts->tv_nsec);
		return nsec >= add_ns;
	}
}

static int read_file_trim(const char *path, char *buf, size_t len)
{
	FILE *f;
	size_t n;

	f = fopen(path, "r");
	if (!f)
		return -1;
	if (!fgets(buf, (int)len, f)) {
		fclose(f);
		return -1;
	}
	fclose(f);
	n = strlen(buf);
	while (n > 0 && (buf[n - 1] == '\n' || buf[n - 1] == '\r' || buf[n - 1] == ' '))
		buf[--n] = '\0';
	return 0;
}

static int read_report_len(bool multi_mode)
{
	char buf[32];

	if (read_file_trim(REPORT_LEN_PATH, buf, sizeof(buf)) == 0)
		return atoi(buf);
	return multi_mode ? 64 : 8;
}

static bool host_configured(void)
{
	char udc[128];
	char state_path[256];
	char state[32];

	if (read_file_trim(GADGET_UDC_PATH, udc, sizeof(udc)) != 0 || udc[0] == '\0')
		return false;

	snprintf(state_path, sizeof(state_path), "/sys/class/udc/%s/state", udc);
	if (read_file_trim(state_path, state, sizeof(state)) != 0)
		return false;

	return strcmp(state, "configured") == 0;
}

static long long timespec_elapsed_ns(const struct timespec *start,
				     const struct timespec *now)
{
	return (long long)(now->tv_sec - start->tv_sec) * 1000000000LL +
	       (long long)(now->tv_nsec - start->tv_nsec);
}

static bool wait_for_host(void)
{
	struct timespec start, now;
	const long long wait_ns = (long long)HOST_WAIT_SEC * 1000000000LL;

	clock_gettime(CLOCK_MONOTONIC, &start);
	for (;;) {
		if (host_configured())
			return true;
		clock_gettime(CLOCK_MONOTONIC, &now);
		if (timespec_elapsed_ns(&start, &now) >= wait_ns)
			break;
		usleep(100000);
	}
	return false;
}

/*
 * 单点模式：把 (tip, 本地 x/y) 转成 report_length 字节的 HID 报告
 * tip=false 时仍写出“抬起”报告（坐标可为 0），主机才能看到松手
 */
static void build_report_single(struct app *a, uint8_t *buf, bool tip, int x, int y)
{
	int hx = 0, hy = 0;

	memset(buf, 0, (size_t)a->report_len);
	if (tip) {
		/* 仅在有触摸时做坐标映射 */
		hx = scale_coord(x, SRC_X_MAX, HID_X_MAX);
		hy = scale_coord(y, SRC_Y_MAX, HID_Y_MAX);
#if INVERT_Y
		hy = HID_Y_MAX - hy;
#endif
	}

	/* 8 字节单点报告（与 usb-gadget-touch.sh 默认描述符一致） */
	buf[0] = 0x01;				/* Report ID */
	buf[1] = tip ? 0x01 : 0x00;		/* Tip Switch */
	buf[2] = 0x00;				/* Contact Identifier */
	if (tip) {
		buf[3] = (uint8_t)(hx & 0xff);
		buf[4] = (uint8_t)(hx >> 8);
		buf[5] = (uint8_t)(hy & 0xff);
		buf[6] = (uint8_t)(hy >> 8);
	}
	buf[7] = tip ? 0x01 : 0x00;		/* Contact Count */
}

/*
 * 多点模式：5 个 slot 各映射为 6 字节，格式与单指在描述符里的一段相同
 *   [flags 0x03][X lo][X hi][Y lo][Y hi][contact_id]
 * byte30 = 当前按下手指数
 */
static void build_report_multi(struct app *a, uint8_t *buf)
{
	int i, count = 0;

	memset(buf, 0, (size_t)a->report_len);
	for (i = 0; i < NUM_SLOTS; i++) {
		int off = i * 6;
		int hx, hy;

		if (!a->slots[i].active)
			continue;
		count++;
		buf[off] = 0x03;
		hx = scale_coord(a->slots[i].x, SRC_X_MAX, HID_X_MAX);
		hy = scale_coord(a->slots[i].y, SRC_Y_MAX, HID_Y_MAX);
#if INVERT_Y
		hy = HID_Y_MAX - hy;
#endif
		buf[off + 1] = (uint8_t)(hx & 0xff);
		buf[off + 2] = (uint8_t)(hx >> 8);
		buf[off + 3] = (uint8_t)(hy & 0xff);
		buf[off + 4] = (uint8_t)(hy >> 8);
		buf[off + 5] = (uint8_t)i;	/* Contact ID = slot 号 */
	}
	if (a->report_len > 30)
		buf[30] = (uint8_t)count;
}

/*
 * 根据 slots[] 生成一整份待写入 hidg 的报告
 * 单点：合并 BTN_TOUCH + 第一个 active 手指 -> build_report_single
 */
static void build_report(struct app *a, uint8_t *buf)
{
	int i;
	bool tip = a->btn_touch;
	int x = 0, y = 0;

	if (a->multi_mode) {
		build_report_multi(a, buf);
		return;
	}

	for (i = 0; i < NUM_SLOTS; i++) {
		if (a->slots[i].active) {
			tip = true;
			x = a->slots[i].x;
			y = a->slots[i].y;
			break;	/* 单点只转发第一根手指 */
		}
	}
	build_report_single(a, buf, tip, x, y);
}

static bool report_equal(const uint8_t *a, const uint8_t *b, int len)
{
	return memcmp(a, b, (size_t)len) == 0;
}

static int hidg_open(struct app *a)
{
	if (a->hid_fd >= 0) {
		close(a->hid_fd);
		a->hid_fd = -1;
	}
	a->hid_fd = open(HIDG_DEV_DEFAULT, O_WRONLY);
	return a->hid_fd;
}

static bool hidg_write(struct app *a, const uint8_t *report)
{
	ssize_t n;

	if (!host_configured())
		return false;

	if (!timespec_cmp_now(&a->last_write_ts, MIN_WRITE_INTERVAL_NS))
		return false;

	n = write(a->hid_fd, report, (size_t)a->report_len);
	if (n != a->report_len) {
		if (n < 0 && (errno == EPIPE || errno == ESHUTDOWN ||
			      errno == ECONNRESET || errno == 108)) {
			usleep(300000);
			if (host_configured() && hidg_open(a) >= 0)
				return false;
		}
		return false;
	}

	clock_gettime(CLOCK_MONOTONIC, &a->last_write_ts);
	return true;
}

/*
 * 一帧 input 结束（SYN_REPORT）：此时 slots[] 已更新完，才组 HID 并 write
 * 这是 event0 -> hidg 转换的“输出触发点”
 */
static void handle_syn_report(struct app *a)
{
	uint8_t report[REPORT_MAX];

	if (!host_configured())
		return;

	build_report(a, report);
	/* 与上一包相同则跳过，避免空闲 SYN 刷屏导致 USB 端点错误 */
	if (a->have_last && report_equal(report, a->last_report, a->report_len))
		return;

	if (hidg_write(a, report)) {
		memcpy(a->last_report, report, (size_t)a->report_len);
		a->have_last = true;
	}
}

/*
 * 解析单条 input_event，更新 slots / btn_touch；不在此处写 hidg
 *
 * Goodix Type B 多点典型一帧顺序示例:
 *   SLOT=0 -> TRACKING_ID=0 -> POSITION_X -> POSITION_Y -> SYN_REPORT
 * 抬起: TRACKING_ID=-1 -> SYN_REPORT
 */
static void handle_input_event(struct app *a, const struct input_event *ev)
{
	if (ev->type == EV_ABS) {
		switch (ev->code) {
		case ABS_MT_SLOT:
			a->cur_slot = ev->value;
			break;
		case ABS_MT_TRACKING_ID:
			if (a->cur_slot >= 0 && a->cur_slot < NUM_SLOTS)
				a->slots[a->cur_slot].active = ev->value >= 0;
			break;
		case ABS_MT_POSITION_X:
			if (a->cur_slot >= 0 && a->cur_slot < NUM_SLOTS)
				a->slots[a->cur_slot].x = ev->value;
			break;
		case ABS_MT_POSITION_Y:
			if (a->cur_slot >= 0 && a->cur_slot < NUM_SLOTS)
				a->slots[a->cur_slot].y = ev->value;
			break;
		case ABS_X:
			a->slots[0].x = ev->value;
			break;
		case ABS_Y:
			a->slots[0].y = ev->value;
			break;
		default:
			break;
		}
	} else if (ev->type == EV_KEY && ev->code == BTN_TOUCH) {
		a->btn_touch = ev->value != 0;
	} else if (ev->type == EV_SYN && ev->code == SYN_REPORT) {
		handle_syn_report(a);
	}
}

int main(int argc, char **argv)
{
	struct app a;
	const char *mode;
	struct pollfd pfd;
	struct input_event ev;
	int ret;

	(void)argc;
	(void)argv;

	memset(&a, 0, sizeof(a));
	a.ev_fd = -1;
	a.hid_fd = -1;
	a.cur_slot = 0;

	mode = getenv("TOUCH_MODE");
	a.multi_mode = mode && strcmp(mode, "multi") == 0;

	a.report_len = read_report_len(a.multi_mode);
	printf("mode=%s report_length=%d\n",
	       a.multi_mode ? "multi" : "single", a.report_len);
	fflush(stdout);

	if (!wait_for_host())
		fprintf(stderr, "warn: USB host not configured; touch after host enumerates\n");

	a.ev_fd = open(INPUT_DEV_DEFAULT, O_RDONLY | O_CLOEXEC);
	if (a.ev_fd < 0) {
		perror("open " INPUT_DEV_DEFAULT);
		return 1;
	}

	if (hidg_open(&a) < 0) {
		perror("open " HIDG_DEV_DEFAULT);
		close(a.ev_fd);
		return 1;
	}

#if GRAB_INPUT
	if (ioctl(a.ev_fd, EVIOCGRAB, (void *)1) < 0)
		perror("EVIOCGRAB");
#endif

	memset(&a.last_write_ts, 0, sizeof(a.last_write_ts));

	pfd.fd = a.ev_fd;
	pfd.events = POLLIN;

	while (1) {
		ret = poll(&pfd, 1, POLL_TIMEOUT_MS);
		if (ret < 0) {
			if (errno == EINTR)
				continue;
			perror("poll");
			break;
		}
		if (ret == 0)
			continue;

		/* 读尽当前可读的所有 input_event，直到 SYN_REPORT 触发转换 */
		while (read(a.ev_fd, &ev, sizeof(ev)) == sizeof(ev))
			handle_input_event(&a, &ev);
	}

	close(a.hid_fd);
	close(a.ev_fd);
	return 0;
}

File: otg-touch-screen/touch-to-hidg.py
#!/usr/bin/env python3
# 使用 Python3 运行；板子上需有 python3

"""
将板载触摸屏 (/dev/input/event0) 转发到 USB HID Gadget (/dev/hidg0)。

数据流: Goodix 本地屏 -> Linux input 子系统 -> 本程序 -> hidg0 -> USB 主机

默认单点 (TOUCH_MODE=single)；报告长度 8 字节，与 usb-gadget-touch.sh 一致。
多点: TOUCH_MODE=multi，gadget 须为 5 点描述符且 report_length=64。
"""

import errno
# 标准 errno 常量，用于判断 USB 写失败类型

import fcntl
# ioctl，用于 EVIOCGRAB 独占本地触摸屏

import os
# 打开 hidg、读 configfs、os.write

import struct
# 解析 input_event、打包 HID 坐标 (uint16 LE)

import select
# 非阻塞等待 /dev/input/event0 可读

import time
# monotonic 计时：限频、等待主机枚举

# --- Linux input 事件类型与常用码 (linux/input-event-codes.h) ---
EV_SYN = 0x00
# 同步类事件

EV_KEY = 0x01
# 按键类（含 BTN_TOUCH）

EV_ABS = 0x03
# 绝对坐标类（触摸 X/Y、多点槽位等）

SYN_REPORT = 0
# 标志本帧 input 事件结束，应在此后发送 HID 报告

ABS_X = 0x00
# 单点绝对 X（部分驱动在 SYN 时同步给出）

ABS_Y = 0x01
# 单点绝对 Y

ABS_MT_SLOT = 0x2F
# 多点协议：当前正在更新哪个 slot (0~4)

ABS_MT_TRACKING_ID = 0x39
# 多点：>=0 表示按下并分配 tracking id，-1 表示该 slot 抬起

ABS_MT_POSITION_X = 0x35
# 多点：当前 slot 的 X

ABS_MT_POSITION_Y = 0x36
# 多点：当前 slot 的 Y

BTN_TOUCH = 0x14A
# 单点触摸按下(1)/抬起(0)，作 build_report 的辅助

# --- 设备路径与运行参数 ---
INPUT_DEV = "/dev/input/event0"
# 板载 Goodix 电容屏对应的 input 节点

HIDG_DEV = "/dev/hidg0"
# USB gadget HID 字符设备；写入字节数须等于 report_length

GADGET_DIR = "/sys/kernel/config/usb_gadget/touch"
# gadget 的 configfs 路径，用于读 UDC 状态与 report_length

TOUCH_MODE = os.environ.get("TOUCH_MODE", "single").lower()
# 环境变量 TOUCH_MODE：single（默认）或 multi

NUM_SLOTS = 5
# 与 Goodix 最多 5 点一致；单点模式只取第一个 active 触点

SRC_X_MAX = 1023
# 本地触摸屏 X 逻辑最大值（evtest 中 Max）

SRC_Y_MAX = 599
# 本地触摸屏 Y 逻辑最大值

HID_X_MAX = 4095
# HID 描述符中 X 逻辑最大值

HID_Y_MAX = 4095
# HID 描述符中 Y 逻辑最大值

INVERT_Y = False
# True 时 Y 轴翻转（部分 Windows 坐标系上下颠倒）

GRAB_INPUT = True
# True：独占 event0，触摸只发 USB，板子 UI 不再响应

HOST_WAIT_SEC = 15
# 启动后等待 USB 主机进入 configured 的最长秒数

MIN_WRITE_INTERVAL = 0.008
# 两次写 hidg 最小间隔（秒），降低 STM32 dwc2 端点错误风险

SELECT_TIMEOUT = 1.0
# select 超时，便于响应 Ctrl+C

EVENT_FMT = "llHHi"
# struct input_event: time.tv_sec, time.tv_usec, type, code, value

EVENT_SIZE = struct.calcsize(EVENT_FMT)
# 单个 input_event 结构体字节数（通常 24）


def read_report_len():
    """从 configfs 读取 report_length；每次 write(hidg) 必须写满该长度。"""
    path = os.path.join(GADGET_DIR, "functions/hid.usb0/report_length")
    try:
        with open(path, "r") as f:
            return int(f.read().strip())
    except (OSError, ValueError):
        return 8 if TOUCH_MODE == "single" else 64
        # 读失败时的默认值：单点 8(Win10)，多点 64


REPORT_LEN = read_report_len()
# 模块加载时读一次；main() 中会再次刷新


def scale(v, src_max, dst_max):
    """将本地坐标 [0,src_max] 线性映射到 [0,dst_max]。"""
    v = max(0, min(int(v), src_max))
    return int(v * dst_max / src_max)


def host_configured():
    """读 UDC state；为 configured 时表示 USB 主机已完成枚举，可以写 hidg。"""
    try:
        with open(os.path.join(GADGET_DIR, "UDC"), "r") as f:
            udc = f.read().strip()
        if not udc:
            return False
        with open(f"/sys/class/udc/{udc}/state", "r") as f:
            return f.read().strip() == "configured"
    except OSError:
        return False


def wait_for_host():
    """启动时轮询等待主机 configured，避免过早写 hidg 导致 -108/断线。"""
    deadline = time.monotonic() + HOST_WAIT_SEC
    while time.monotonic() < deadline:
        if host_configured():
            return True
        time.sleep(0.1)
    return False


class Slot:
    """Type B 多点协议中一个 slot 的状态。"""

    __slots__ = ("active", "x", "y")
    # 限制属性，节省内存

    def __init__(self):
        self.active = False
        # 该 slot 是否有手指（TRACKING_ID >= 0）
        self.x = 0
        self.y = 0
        # 当前坐标（本地分辨率）


def build_report_single(slots, btn_touch=False):
    """组单点 HID 报告：8 字节，含 Report ID（与 usb-gadget-touch.sh 默认描述符一致）。"""
    tip = btn_touch
    x = y = 0
    for s in slots:
        if s.active:
            tip = True
            x, y = s.x, s.y
            break
    hx = scale(x, SRC_X_MAX, HID_X_MAX) if tip else 0
    hy = scale(y, SRC_Y_MAX, HID_Y_MAX) if tip else 0
    if tip and INVERT_Y:
        hy = HID_Y_MAX - hy

    buf = bytearray(REPORT_LEN)
    buf[0] = 0x01
    buf[1] = 0x01 if tip else 0x00
    buf[2] = 0x00
    if tip:
        struct.pack_into("<HH", buf, 3, hx, hy)
    buf[7] = 0x01 if tip else 0x00
    return bytes(buf)


def build_report_multi(slots):
    """
    5 点 HID 报告：每指 6 字节，byte30 为 contact count。
    仅 TOUCH_MODE=multi 且 gadget report_length=64 时使用。
    """
    buf = bytearray(REPORT_LEN)
    count = 0
    for i, s in enumerate(slots):
        off = i * 6
        if s.active:
            count += 1
            buf[off] = 0x03
            x = scale(s.x, SRC_X_MAX, HID_X_MAX)
            y = scale(s.y, SRC_Y_MAX, HID_Y_MAX)
            if INVERT_Y:
                y = HID_Y_MAX - y
            struct.pack_into("<HH", buf, off + 1, x, y)
            buf[off + 5] = i
    if REPORT_LEN > 30:
        buf[30] = count
    return bytes(buf)


def build_report(slots, btn_touch=False):
    if TOUCH_MODE == "multi":
        return build_report_multi(slots)
    return build_report_single(slots, btn_touch)


class HidgWriter:
    """封装对 /dev/hidg0 的写操作：检查 configured、限速、断线重开。"""

    def __init__(self, path):
        self.path = path
        self.fd = None
        self.last_write = 0.0
        self.open()

    def open(self):
        if self.fd is not None:
            try:
                os.close(self.fd)
            except OSError:
                pass
        self.fd = os.open(self.path, os.O_WRONLY)

    def write(self, report):
        if not host_configured():
            return False
        now = time.monotonic()
        if now - self.last_write < MIN_WRITE_INTERVAL:
            return False
        try:
            n = os.write(self.fd, report)
            if n != len(report):
                raise OSError("short write", n, len(report))
            self.last_write = now
            return True
        except OSError as e:
            if e.errno in (errno.EPIPE, errno.ESHUTDOWN, errno.ECONNRESET, 108):
                time.sleep(0.3)
                if host_configured():
                    try:
                        self.open()
                    except OSError:
                        pass
            return False


def main():
    global REPORT_LEN
    REPORT_LEN = read_report_len()
    print("mode=%s report_length=%d" % (TOUCH_MODE, REPORT_LEN), flush=True)

    if not wait_for_host():
        print("warn: USB host not configured; touch after Windows enumerates", flush=True)

    ev = open(INPUT_DEV, "rb")
    hid = HidgWriter(HIDG_DEV)

    if GRAB_INPUT:
        fcntl.ioctl(ev, 0x40044590, 1)
        # EVIOCGRAB：独占触摸屏

    slots = [Slot() for _ in range(NUM_SLOTS)]
    cur_slot = 0
    btn_touch = False
    last_report = None

    while True:
        try:
            r, _, _ = select.select([ev], [], [], SELECT_TIMEOUT)
        except (KeyboardInterrupt, OSError):
            break
        if not r:
            continue
        data = ev.read(EVENT_SIZE)
        if not data or len(data) < EVENT_SIZE:
            continue
        _, _, typ, code, val = struct.unpack(EVENT_FMT, data)

        if typ == EV_ABS:
            if code == ABS_MT_SLOT:
                cur_slot = val
            elif code == ABS_MT_TRACKING_ID:
                if 0 <= cur_slot < NUM_SLOTS:
                    slots[cur_slot].active = val >= 0
            elif code in (ABS_MT_POSITION_X, ABS_X):
                if code == ABS_MT_POSITION_X and 0 <= cur_slot < NUM_SLOTS:
                    slots[cur_slot].x = val
                elif code == ABS_X:
                    slots[0].x = val
            elif code in (ABS_MT_POSITION_Y, ABS_Y):
                if code == ABS_MT_POSITION_Y and 0 <= cur_slot < NUM_SLOTS:
                    slots[cur_slot].y = val
                elif code == ABS_Y:
                    slots[0].y = val
        elif typ == EV_KEY and code == BTN_TOUCH:
            btn_touch = bool(val)
        elif typ == EV_SYN and code == SYN_REPORT:
            if not host_configured():
                continue
            report = build_report(slots, btn_touch)
            if report == last_report:
                continue
            if hid.write(report):
                last_report = report

    try:
        os.close(hid.fd)
        ev.close()
    except OSError:
        pass


if __name__ == "__main__":
    main()
